Initial implementation of Throttle middleware

keirlawson opened the pull request:
Feedback appreciated!

There is definitely room for more features, in a subsequent PR I'd like to make taking tokens a function of the `Request` so that throttling can be based on request sizes or IPs rather than just simply the number of requests.  I'd also like to enable customisation of the response to throttled requests.  If these are considered essential then I can add them to this PR however my preference would be to get this merged before iterating further.

jcranky commented on server/src/main/scala/org/http4s/server/middleware/Throttle.scala:
```diff
@@ -0,0 +1,90 @@
+import cats.implicits._
+import fs2.Stream
+
+sealed trait TokenAvailability
```
`Token` makes my mind think about authentication / Authorization. Is this a common terminology used for throttling as well?

SystemFw replied:
It is. The algorithm Keir used here is commonly known as a "Token bucket". 
If you think about it _is_ similar to authentication, in both cases a token grants you the capability to do something.

jcranky replied:
@SystemFw fair, thanks for the clarification :)

rossabaker commented on server/src/main/scala/org/http4s/server/middleware/Throttle.scala:
```diff
@@ -0,0 +1,90 @@
+import cats.implicits._
+import fs2.Stream
+
+sealed trait TokenAvailability
```
`sealed abstract class TokenAvailability extends Product with Serializable`: noisier, but better inference an bin compat.

This is dumping a bunch of symbols into the middleware package.  I wonder whether this ADT, and the `TokenBucket`, should be moved inside `Throttle`.  Or maybe, in the opposite direction, it could be useful on a `Client`, and we should move it into `org.http4s.util`.  :thinking:

SystemFw replied:
Slight preference for having it into `Throttle` and not `util` for the time being.

rossabaker replied:
But move `Throttle` to another package or leave it here?  I agree with moving the ADT into `Throttle` regardless.

`org.http4s.utill` is a cursed place of "things that make us wish that Scala were more amenable to microlibraries," so I do acknowledge the disgust.

SystemFw replied:
I won't oppose moving into `util` once we have the need. For now I'd rather keep it here so that the code is easier to follow. 

On a related not wrt reusability, client-side limiting imho is not a great fit for a token bucket. `upperbound` (which focuses on that) has a different design for that reason

rossabaker commented on server/src/main/scala/org/http4s/server/middleware/Throttle.scala:
```diff
@@ -0,0 +1,90 @@
+      implicit F: Concurrent[F],
+      timer: Timer[F]): Stream[F, TokenBucket[F]] =
+    Stream.eval(Ref[F].of(capacity)).flatMap { counter =>
+      val refill = Stream
```
I've implemented this previously by storing the last time a token was requested.  Then you can interpolate out how much has been refilled from last time before deciding whether or not to grant a token, and not need a background process.  This approach results in fractional tokens.

I think we'd also then get away with an `F[TokenBucket[F]]` instead of a `Stream[F, TokenBucket[F]]`.

SystemFw replied:
> I think we'd also then get away with an F[TokenBucket[F]] instead of a Stream[F, TokenBucket[F]].

we can do this regardless btw. Returning a Resource[TokenBucket]` by `compile.drain.start` and `cancel` as the shutdown action.

You mentioned the refill approach a couple times but never with enough detail to understand _how_ you did it, so this is interesting :)

SystemFw replied:
Just out of curiosity, it wouldn't work with a TokenBucket that had the task waiting rather than rejected, right? (if more recent requests don't arrive, old tasks are stuck)

rossabaker replied:
I think my rejection message included a "retry-after" like value.  I don't think I maintained a queue.  In either case, we can compute the time until full and try again.  It does get messy if we're returning the same retry-after to multiple requestors and get slammed.

rossabaker commented on server/src/main/scala/org/http4s/server/middleware/Throttle.scala:
```diff
@@ -0,0 +1,90 @@
+    * @param http the service to transform.
+    * @return a singleton stream containing the transformed service.
+    */
+  def apply[F[_], G[_]](bucket: TokenBucket[F])(http: Http[F, G])(
```
Should we make the 429 response customizable?

rossabaker replied:
Now that I mention Retry-After above, hey, how about a Retry-After? :smile:

keirlawson replied:
As I said in the PR message I want to implement this, I was just planning to do so in a follow-up PR, but if the feeling is that it is essential I can do it in this one.

SystemFw replied:
I think customisable responses are really easy to add here, and perhaps the rest can go in another PR?

keirlawson commented:
Rebased on latest master, swapped to @rossabaker's solution for refilling bucket, added response customisation and retry-after set to when the bucket ought to be full.  

I'm not totally convinced with the retry-after value, you could argue it should be when the next token will be available rather than when the bucket should be full, though in the real world if clients pay attention to that things will likely blow up, seems to me neither the next-token-time nor the bucket-full-time really make sense and it would be best to leave it up to the user as I think any value is either arbitrary or app-specific.

SystemFw commented on server/src/main/scala/org/http4s/server/middleware/Throttle.scala:
```diff
@@ -0,0 +1,120 @@
+                } else {
+                  val timeToFull = (refillEvery.toNanos * capacity) - timeDifference
+                  val successResponse = TokenUnavailable(Some(FiniteDuration(timeToFull, NANOSECONDS)))
+                  setter((newTokenTotal, currentTime)).map(success => if (success) Some(successResponse) else None)
```
`TokenUnavailable(Some(FiniteDuration(timeToFull, NANOSECONDS)))`

apologies for the bikeshedding (feel free to ignore)

`TokenUnavailable(timeToFull.nanos.some)`

keirlawson replied:
Bikeshedding appreciated, I had the feeling there would be a more concise way to do this...

SystemFw commented on server/src/main/scala/org/http4s/server/middleware/Throttle.scala:
```diff
@@ -0,0 +1,120 @@
+
+  sealed abstract class TokenAvailability extends Product with Serializable
+  case class TokenAvailable() extends TokenAvailability
+  case class TokenUnavailable(retryAfter: Option[FiniteDuration]) extends TokenAvailability
```
Is this `Option` even `None`?

keirlawson replied:
It could be with other `TokenBucket` impls, ie I envisage an "outstanding requests" version where you issue tokens when there are less than n requests currently being executed, in that scenario there is no obvious time-to-full

rossabaker commented on server/src/main/scala/org/http4s/server/middleware/Throttle.scala:
```diff
@@ -1,90 +1,125 @@
 
-object TokenBucket {
+  sealed abstract class TokenAvailability extends Product with Serializable
+  case class TokenAvailable() extends TokenAvailability
```
Why did this cease to be a case object?

rossabaker commented on server/src/main/scala/org/http4s/server/middleware/Throttle.scala:
```diff
@@ -1,90 +1,125 @@
-object TokenBucket {
+  sealed abstract class TokenAvailability extends Product with Serializable
+  case class TokenAvailable() extends TokenAvailability
+  case class TokenUnavailable(retryAfter: Option[FiniteDuration]) extends TokenAvailability
```
Any reason not to make this a final case class?

rossabaker commented on server/src/main/scala/org/http4s/server/middleware/Throttle.scala:
```diff
@@ -1,90 +1,125 @@
-      implicit F: Concurrent[F]): Http[F, G] =
+  def apply[F[_], G[_]](
+      bucket: TokenBucket[F],
+      throttleResponse: Option[FiniteDuration] => Response[G] = defaultResponse[G] _)(
```
I like this, because people can alternately fill in a timestamp instead of a duration if they prefer.  (Or of course add a custom body.)

rossabaker commented on server/src/main/scala/org/http4s/server/middleware/Throttle.scala:
```diff
@@ -1,90 +1,125 @@
     */
   def apply[F[_], G[_]](amount: Int, per: FiniteDuration)(
-      http: Http[F, G])(implicit F: Concurrent[F], timer: Timer[F]): Stream[F, Http[F, G]] = {
+      http: Http[F, G])(implicit F: Sync[F], timer: Clock[F]): F[Http[F, G]] = {
```
This is kind of confusing: the types are the same as `TokenBucket.local`, but we do extra math first.

I wonder whether this overload carries its weight at all, given that it doesn't allow customization of the throttled response.  I imagine we could live without it, but I'd defer to you since you've probably tried to use this and I haven't.

keirlawson replied:
I'm not totally attached but my thinking was that some people will not care about the token bucket impl so easier to have a default where you don't need to construct your own, and the x requests per y time was a more "natural" way that people would think about their rate limit, ie "100 per second" rather than "1 per 0.6 seconds"

SystemFw replied:
I'm very slightly 👎 on the overload

keirlawson commented:
Per discussion in the Gitter channel, have changed behavior to not use the retry-after header by default, however time-to-next-token value is made available to the `throttleRepose` function should someone wish to use it, time-to-bucket-full should be derivable from this + refill rate.

jmcardon commented on server/src/main/scala/org/http4s/server/middleware/Throttle.scala:
```diff
@@ -0,0 +1,121 @@
+              case ((previousTokens, previousTime), setter) =>
+                getTime.flatMap(currentTime => {
+                  val timeDifference = currentTime - previousTime
+                  val tokensToAdd = timeDifference.toDouble / refillEvery.toNanos.toDouble
```
Why are you doing a `toDouble` conversion just for the sake of using `Math.min`?

keirlawson replied:
`tokensToAdd` needs to allow for fractional tokens, if I remove both `toDouble`s from that line it will wind up a `long`

jmcardon commented on server/src/main/scala/org/http4s/server/middleware/Throttle.scala:
```diff
@@ -0,0 +1,121 @@
+        clock: Clock[F]): F[TokenBucket[F]] = {
+
+      def getTime = clock.monotonic(NANOSECONDS)
+      val bucket = getTime.flatMap(time => Ref[F].of((capacity.toDouble, time)))
```
why is this a `Double` over a `Long`? You'll essentially be working with integral amounts most of the time.

keirlawson replied:
If we want to allow fractional tokens, which is the basis of the algorithm suggested by @rossabaker then I think we need a double? Unless I am missing something?

jmcardon commented on server/src/main/scala/org/http4s/server/middleware/Throttle.scala:
```diff
@@ -0,0 +1,121 @@
+                  } else {
+                    val timeToNextToken = refillEvery.toNanos - timeDifference
+                    val successResponse = TokenUnavailable(timeToNextToken.nanos.some)
+                    setter((newTokenTotal, currentTime)).map(_.guard[Option].as(successResponse))
```
Just to double check: Is this intended that if you keep sending requests while being throttled, it will remain throttled? I.e: If this is the case and you hit a throttle, while continuing to send requests, it will not let any through after the throttle, since it always updated the current time here.

keirlawson replied:
That is not the intention, and I don't *think* that is the behaviour here, though my reasoning could be wrong... it should keep adding fractional tokens until there is a whole token at which point the request passes, if you're unconvinced then I can add a test that with (hopefully!) prove this?

jmcardon replied:
Yeah I was wrong, didn't take into account fractional updates.

jmcardon commented on server/src/test/scala/org/http4s/server/middleware/ThrottleSpec.scala:
```diff
@@ -0,0 +1,143 @@
+        val takeFiveTokens: IO[List[TokenAvailability]] = (1 to 5).toList.traverse(_ => {
+          testee.takeToken
+        })
+        testTimer.sleep(300.milliseconds) >> takeFiveTokens >> testee.takeToken
```
why does it sleep 300 millis, take 5 tokens then take another token? 

I might be misreading: This test aims to make sure that if trying to take a token at capacity, it won't add another token until demand goes down?

keirlawson replied:
No, the intention of the test was to ensure the bucket never adds tokens beyond its capacity, ie we set it to capacity, wait a few refill intervals, then check that there are still only `capacity` tokens.

jmcardon replied:
My mistake :+1:
